Assets/MixedRealityToolkit/_Core/Definitions/Utilities/InscribedRectangle.cs

```diff
@@ -0,0 +1,570 @@
+        /// vary the size, until the largest fit-able rectangle is found.
+        /// This is then repeated for predefined angles (0-180 in steps of 15)
+        /// and aspect ratios (1 to 15 in steps of 0.5).
+        /// </summary>
+        /// <param name="geometryEdges">The boundary geometry.</param>
+        /// <param name="randomSeed">Random number generator seed.</param>
+        /// <remarks>
+        /// For the most reproducable results, use the same randomSeed value 
+        /// each time this method is called.
+        /// </remarks>
+        private void FindInscribedRectangle(Edge[] geometryEdges, int randomSeed)
+        {
+            // Clear previous rectangle
+            Center = EdgeUtilities.InvalidPoint;
+            Width = 0;
+            Height = 0;
+            Angle = 0;
+
+            float minX = EdgeUtilities.maxWidth;
+            float minY = EdgeUtilities.maxWidth;
+            float maxX = -EdgeUtilities.maxWidth;
+            float maxY = -EdgeUtilities.maxWidth;
+
+            // Find min x, min y, max x, max y 
+            for (int i = 0; i < geometryEdges.Length; i++)
+            {
+                Edge edge = geometryEdges[i];
+
+                if ((edge.PointA.x < minX) || (edge.PointB.x < minX))
+                {
+                    minX = Mathf.Min(edge.PointA.x, edge.PointB.x);
+                }
+
+                if ((edge.PointA.y < minY) || (edge.PointB.y < minY))
+                {
+                    minY = Mathf.Min(edge.PointA.y, edge.PointB.y);
+                }
+
+                if ((edge.PointA.x > maxX) || (edge.PointB.x > maxX))
+                {
+                    maxX = Mathf.Max(edge.PointA.x, edge.PointB.x);
+                }
+
+                if ((edge.PointA.y > maxY) || (edge.PointB.y > maxY))
+                {
+                    maxY = Mathf.Min(edge.PointA.y, edge.PointB.y);
+                }
+
+            }
+
+            // Generate random points until we have randomPointCount starting points
+            Vector2[] startingPoints = new Vector2[randomPointCount];
+            {
+                System.Random random = new System.Random(randomSeed);
+                for (int i = 0; i < startingPoints.Length; i++)
+                {
+                    Vector2 candidatePoint;
+
+                    do
+                    {
+                        candidatePoint.x = ((float)random.NextDouble() * (maxX - minX)) + minX;
+                        candidatePoint.y = ((float)random.NextDouble() * (maxY - minY)) + minY;
+                    }
+                    while (!EdgeUtilities.IsInsideBoundary(geometryEdges, candidatePoint));
+
+                    startingPoints[i] = candidatePoint;
+                }
+
+            }
+
+            // Angles to use for fitting the rectangle within the boundary.
+            float[] angles = { 0, 15, 30, 45, 60, 75, 90, 105, 120, 135, 150, 165 };
```

Comment:
couldn't we just declare this at the class scope so we don't allocate every time we look for  a new rectangle?

Reply:
this was how the original code was written, i didnt want to tamper too much with it. easy enough chanbge